### DISCUSSION TITLE
Return pytrees without implicit copy

### DISCUSSION THREAD

**Discussion by dionhaefner** [jax-ml/jax#6237] [Q&A] [answered]
Suppose I have a large `state` tuple that I pass to jitted functions, like this:

```python
from collections import namedtuple
import jax
import jax.numpy as jnp

# state contains 100 different JAX arrays
State = namedtuple("State", ["v{i}" for i in range(100)])

@jax.jit
def foo(state):
    new_v2 = state.v0 + state.v1
    return state._replace(v2=new_v2)

state = State(*[jnp.ones(1000) * i for i in range(100)])
new_state = foo(state)
```

It is my understanding that the state object returned by `foo` contains copies of all 100 JAX arrays, even if they are unchanged. I.e., `id(new_state.v0) != id(state.v0)`, even though `v0` wasn't touched by `foo`.

Is there a way to prevent this, or is this behavior just baked into how JAX handles pytrees?

**Comment by davisyoshida** [comment]
You can use `jax.jit`'s `donate_argnums` argument if you're planning on discarding the old state. That would allow to re-use the buffers for the unmodified arrays and hopefully it would recognize that leaving the arrays in place is more efficient (`jit` is a magical black box to me so I'm not sure).

I'm also not sure if the IDs not being equal even implies that there was a copy on the GPU, since they could just be two different references to the same location in memory (I'm out of my depth here).

**Reply by dionhaefner** [reply]
Thanks for the pointer. I tried `donate_argnums`, but got this:

```
UserWarning: Some donated buffers were not usable: f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}, f32[100000]{0}
```

So it seems like it doesn't reuse any of the buffers.

I am pretty sure that a copy is happening behind the scenes (I checked the memory pointers, and my benchmarks show significant overhead).

**Comment by hawkinsp** [comment, marked as answer]
I think it's a reasonable feature request that JAX should forward buffers without copying them if you do not modify them. We already have logic along those lines for trivial computations, but not for computations that also do some amount of computation.

(It's not really related to pytrees, just to do with how JAX lowers to XLA computations.)

**Reply by mattjj** [reply]
It's not the same thing, but that feels conceptually related to pruning unused inputs (in the sense of not device_putting values that correspond to unused jaxpr inputs), another optimization we want to add / upstream.

**Reply by dionhaefner** [reply]
Wow, I just realized that this suffers from the same problem:

```python
import jax
import jax.numpy as jnp

@jax.jit
def no_op(a, b, c):
    return (a, b, a)

@jax.jit
def some_op(a, b, c):
    return (a, b, a + b)

a, b, c = [jnp.ones(10) for _ in range(3)]
new_a, new_b, new_c = no_op(a, b, c)
print(a.unsafe_buffer_pointer() == new_a.unsafe_buffer_pointer())  # prints True

new_a, new_b, new_c = some_op(a, b, c)
print(a.unsafe_buffer_pointer() == new_a.unsafe_buffer_pointer())  # prints False
```

I'll turn this into an issue, then.

**Reply by dionhaefner** [reply]
To be continued in #6284
